### PR TITLE
Lint unused assoc tys although the trait is used

### PR DESCRIPTION
**Pull request by mu001999**
r? @pnkfelix

### CODE REVIEW

**Comment by mu001999** [comment]
wired.

```
ERROR: Tool `book` was not recorded in tool state.
ERROR: Tool `nomicon` was not recorded in tool state.
ERROR: Tool `reference` was not recorded in tool state.
ERROR: Tool `rust-by-example` was not recorded in tool state.
ERROR: Tool `edition-guide` was not recorded in tool state.
ERROR: Tool `embedded-book` was not recorded in tool state.
```

**Comment by mu001999** [comment]
```
  error: associated type `Bar` is never used
     --> library/core/tests/any.rs:137:14
      |
  136 |     trait Foo {
      |           --- associated type in this trait
  137 |         type Bar;
      |              ^^^
      |
      = note: `-D dead-code` implied by `-D warnings`
      = help: to override `-D warnings` add `#[allow(dead_code)]`
```

the code is:
```rust
#[test]
fn dyn_type_name() {
    trait Foo {
        type Bar;
    }

    assert_eq!(
        "dyn core::ops::function::Fn(i32, i32) -> i32",
        std::any::type_name::<dyn Fn(i32, i32) -> i32>()
    );
    assert_eq!(
        "dyn coretests::any::dyn_type_name::Foo<Bar = i32> \
        + core::marker::Send + core::marker::Sync",
        std::any::type_name::<dyn Foo<Bar = i32> + Send + Sync>()
    );
}
```

seems the assoc type is not marked live if it was used in the trait object.

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -722,6 +784,41 @@ impl<'tcx> Visitor<'tcx> for MarkSymbolVisitor<'tcx> {
 
         self.in_pat = in_pat;
     }
+
+    fn visit_poly_trait_ref(&mut self, t: &'tcx hir::PolyTraitRef<'tcx>) {
+        // mark the assoc type/const appears in poly-trait-ref live
+        if let Some(pathsegment) = t.trait_ref.path.segments.last()
```
Comment (mu001999) [inline compiler/rustc_passes/src/dead.rs:790]: this comes from #127505, and this PR should be merged after that

Comment (trevyn) [reply compiler/rustc_passes/src/dead.rs:808]: Feel free to take it, and the tests etc. from that PR. You've got a better handle on this anyway. :)

**Pull request closed without merging**